core_lib/src/interface/scribblearea.cpp

```diff
@@ -1454,12 +1462,28 @@ void ScribbleArea::toggleOutlines()
     setEffect(SETTING::OUTLINES, mIsSimplified);
 }
 
-void ScribbleArea::toggleShowAllLayers()
+void ScribbleArea::setLayerVisibility(LayerVisibility visibility)
+{
+    mLayerVisibility = visibility;
+    updateAllFrames();
+}
+
+void ScribbleArea::increaseLayerVisibilityIndex()
+{
+    mLayerVisibility = static_cast<LayerVisibility>(static_cast<int>(mLayerVisibility) + 1);
+    if (mLayerVisibility > LayerVisibility::FULL)
```

Comment:
Well I would have converted to ints too, but this isn't quite the way I would go about doing this. In particular there are two issues I have with how this is currently implemented. First, it depends on LayerVisibility::Hidden being the first element and LayerVisibility::FULL being the last element, which is not clear elsewhere and is an odd requirement. Second, mLayerVisibility takes on a value outside of the defined values. To be honest I was surprised this worked at all, but apparently the specifications do allow for this. Still it makes me a little uneasy seeing this and imagining what were to happen if any other part of the program had to deal with these undefined values.

Here's what I would propose instead. Change the enum to this:
```c++
enum LayerVisibility {
    HIDDEN = 0,
    RELATIVE = 1,
    FULL = 2,
    COUNT = 3, // COUNT must always be last
};
```
Then the increaseLayerVisibilityIndex function changes to something like this:
```c++
mLayerVisibility = static_cast<LayerVisibility>((static_cast<int>(mLayerVisibility) + 1) % LayerVisibility::FULL);
updateAllFrames();
```
And decreaseLayerVisibilityIndex would be similar:
```c++
mLayerVisibility = static_cast<LayerVisibility>((static_cast<int>(mLayerVisibility) - 1 + LayerVisibility::FULL) % LayerVisibility::FULL);
updateAllFrames();
```
What do you think?

Reply:
I don't see what you achieve by doing this, if you think it's more safe, I disagree. The only difference here is you use count to iterate because you want to avoid an undefined variable that won't be used anyway and want to be sure the value you increment to matches one of the enums. I don't think is any better than incrementing to an undefined value which right after is reset so that the undefined behaviour will never be used it's untentional.

I've changed it though, adding two operators to increment/decrement within allowed range, no need for count still though.

Reply:
Well I disagree with your disagreeing :yum: . Based on my research ([mostly frome here](https://stackoverflow.com/questions/33812998/is-it-allowed-for-an-enum-to-have-an-unlisted-value#answer-33814842)), incrementing out of the enumeration range results in an unspecified value before C++17 and undefined behavior in C++17. That means the result could technically be any value so the assumption that `FULL+1 > FULL` is not good. When you switched to using a scoped enum this stopped being a problem because the underlying type is set to be int rather than whatever the compiler wants. The overrides are still a nice addition though :+1: 

I still think having COUNT would be useful to avoid accidentally forgetting to change the overrides when adding new enum value, but at least least the changes are close together now. I will add a comment to the code and leave it at that.